internal/connection_manager.go

```diff
@@ -104,35 +105,63 @@ func (cm *connectionManager) connectionClosed(connection *Connection, cause erro
 	}
 }
 
-func (cm *connectionManager) getOrConnect(address core.Address, asOwner bool) (chan *Connection, chan error) {
-	ch := make(chan *Connection, 1)
-	errCh := make(chan error, 1)
-	go func() {
-		//First tries to read under read lock
-		conn := cm.getConnection(address, asOwner)
-		if conn != nil {
-			ch <- conn
-			return
-		}
-		//if not available, checks and creates under write lock
-		newConn, err := cm.getOrConnectInternal(address, asOwner)
-		if err != nil {
-			errCh <- err
-		} else {
-			ch <- newConn
-		}
-	}()
-	return ch, errCh
+func (cm *connectionManagerImpl) getOrTriggerConnect(address core.Address) (*Connection, error) {
+	connection := cm.getConnection(address, false)
+	if connection != nil {
+		return connection, nil
+	}
+	go cm.getOrCreateConnectionInternal(address, false)
+	return nil, core.NewHazelcastIOError("No available connection to address "+address.String(), nil)
+}
+
+func (cm *connectionManagerImpl) getOrConnect(address core.Address, asOwner bool) (*Connection, error) {
+	connection := cm.getConnection(address, asOwner)
+	if connection != nil {
+		return connection, nil
+	}
+	return cm.getOrCreateConnectionInternal(address, asOwner)
+}
+
+func (cm *connectionManagerImpl) getOwnerConnection() *Connection {
+	ownerConnectionAddress := cm.client.ClusterService.getOwnerConnectionAddress()
+	if ownerConnectionAddress == nil {
+		return nil
+	}
+	return cm.getActiveConnection(ownerConnectionAddress)
+}
+
+func (cm *connectionManagerImpl) shutdown() {
+	activeCons := cm.getActiveConnections()
+	for _, con := range activeCons {
+		con.close(core.NewHazelcastClientNotActiveError("client is shutting down", nil))
+	}
 }
 
-func (cm *connectionManager) getConnection(address core.Address, asOwner bool) *Connection {
-	cm.lock.RLock()
-	defer cm.lock.RUnlock()
-	return cm.getConnectionInternal(address, asOwner)
+//internal definitions and methods called inside connection manager process
+
+type connectionManagerImpl struct {
```

Comment:
I think this name does not suit go well. It should be `connectionManager` and there is also an interface named `connectionManager` but as I said earlier, I dont think having an interface and struct in the same package is go way.